library/aesce.c

```diff
@@ -101,66 +101,116 @@ int mbedtls_aesce_has_support(void)
 #endif
 }
 
+/* Single round of AESCE encryption */
+#define AESCE_ENCRYPT_ROUND                   \
+    block = vaeseq_u8(block, vld1q_u8(keys)); \
+    block = vaesmcq_u8(block);                \
+    keys += 16
+/* Two rounds of AESCE encryption */
+#define AESCE_ENCRYPT_ROUND_X2        AESCE_ENCRYPT_ROUND; AESCE_ENCRYPT_ROUND
+
+MBEDTLS_OPTIMIZE_FOR_PERFORMANCE
 static uint8x16_t aesce_encrypt_block(uint8x16_t block,
                                       unsigned char *keys,
                                       int rounds)
 {
-    for (int i = 0; i < rounds - 1; i++) {
-        /* AES AddRoundKey, SubBytes, ShiftRows (in this order).
-         * AddRoundKey adds the round key for the previous round. */
-        block = vaeseq_u8(block, vld1q_u8(keys + i * 16));
-        /* AES mix columns */
-        block = vaesmcq_u8(block);
+    /* Assume either 10, 12 or 14 rounds.
+     * Skip 4 or 2 rounds, if doing 10 or 12 rounds */
+    if (rounds == 10) {
```

Comment:
Did you try with `goto`?

```
    if (rounds > 12) {
        AESCE_ENCRYPT_ROUND_X2;
    }
    if (rounds > 10) {
        AESCE_ENCRYPT_ROUND_X2;
    }
    AESCE_ENCRYPT_ROUND_X2;
    AESCE_ENCRYPT_ROUND_X2;
    AESCE_ENCRYPT_ROUND_X2;
    AESCE_ENCRYPT_ROUND_X2;
```

Reply:
or even with `switch`

```
switch (rounds) {
    case 14:
        AESCE_ENCRYPT_ROUND_X2;
        /* fallthrough */
    case 12:
        AESCE_ENCRYPT_ROUND_X2;
        /* fallthrough */
    case 10:
        AESCE_ENCRYPT_ROUND_X2;
        AESCE_ENCRYPT_ROUND_X2;
        AESCE_ENCRYPT_ROUND_X2;
        AESCE_ENCRYPT_ROUND_X2;
}
```

Reply:
I tried both of those. The variant here was slightly better.